src/core/mesh/qgsmeshtracerenderer.h

```diff
@@ -216,49 +227,243 @@ class QgsMeshStreamField
+    virtual void storeInField( const QPair<QPoint, FieldData> pixelData ) = 0;
+    virtual void initField() = 0;
+    void simplifyChunkTrace( std::list<QPair<QPoint, FieldData>> &shunkTrace );
+
+    virtual bool isTraceExists( const QPoint &pixel ) const = 0;
     bool isTraceOutside( const QPoint &pixel ) const;
 
-    bool filterMag( double value ) const;
+  protected:
 
     QSize mFieldSize;
-    QRect mLayerPixelExtent;
+    QPainter *mPainter = nullptr;
     int mFieldResolution = 1;
+    QPen mPen;
+    QImage mTraceImage;
+
+    QgsMapToPixel mMapToFieldPixel;
+
+  private:
     int mPixelFillingCount = 0;
     int mMaxPixelFillingCount = 0;
-    std::unique_ptr<QgsMeshVectorValueInterpolator> mVectorValueInterpolator;
-    QImage mTraceImage;
-    QVector<bool> mField;
+    std::shared_ptr<QgsMeshVectorValueInterpolator> mVectorValueInterpolator;
     QgsRectangle mLayerExtent;
     QgsRectangle mMapExtent;
-    QgsMapToPixel mMapToFieldPixel;
-    QPen mPen;
-    QPainter *mPainter = nullptr;
     QPoint mFieldTopLeftInDeviceCoordinates;
     bool mValid = false;
     double mMagMax = 0;
     double mPixelFillingDensity;
     double mMinMagFilter = -1;
     double mMaxMagFilter = -1;
-    QgsRenderContext &mRenderContext; //keep the renderer context only to know if the renderer is stopped
+    const QgsRenderContext &mRenderContext; //keep the renderer context only to know if the renderer is stopped
+    bool mMinimizeFieldSize = true; //
 };
 
-class QgsMeshVectorStreamLineRenderer: public QgsMeshVectorRenderer
+/**
+ * \ingroup core
+ *
+ * Class used to draw streamlines from vector field
+ *
+ * \note not available in Python bindings
+ * \since QGIS 3.12
+ */
+class QgsMeshStreamlinesField: public QgsMeshStreamField
+{
+  public:
+    //! Constructor
+    QgsMeshStreamlinesField( const QgsTriangularMesh &triangularMesh,
+                             const QgsMeshDataBlock &datasetVectorValues,
+                             const QgsMeshDataBlock &scalarActiveFaceFlagValues,
+                             const QgsRectangle &layerExtent,
+                             double magMax, bool dataIsOnVertices, QgsRenderContext &rendererContext ):
+      QgsMeshStreamField( triangularMesh, datasetVectorValues, scalarActiveFaceFlagValues, layerExtent, magMax, dataIsOnVertices, rendererContext )
+    {}
+
+  private:
+    void storeInField( const QPair<QPoint, FieldData> pixelData ) override;
+    void initField() override;
+    bool isTraceExists( const QPoint &pixel ) const override;
+    void drawChunkTrace( std::list<QPair<QPoint, FieldData>> &chunkTrace ) override;
+
+    QVector<bool> mField;
+
+};
+
+class QgsMeshParticleTracesField;
+
+/**
+ * \ingroup core
+ *
+ * Used to simulation moving particle
+ *
+ * \note not available in Python bindings
+ * \since QGIS 3.12
+ */
+struct QgsMeshTraceParticle
+{
+  double lifeTime;
+  QPoint position;
+  std::list<QPoint> tail;
```

Comment:
better to use QT classes -> QList

Reply:
QList is not really equivalent with std::list. Here, I need iterator and fast remove and append on the top/back. If Qt class is required, maybe QLinkedList ?

Reply:
better to use qt objects in any case, pick one that suits you